File: src/main/java/com/sourcegraph/cody/chat/BlinkingCursorComponent.java
package com.sourcegraph.cody.chat;

import java.awt.Dimension;
import java.awt.Font;
import java.awt.Graphics;
import javax.swing.JPanel;
import javax.swing.Timer;

public class BlinkingCursorComponent extends JPanel {
  private boolean showCursor;
  private Timer timer;

  public static BlinkingCursorComponent instance = new BlinkingCursorComponent();

  private BlinkingCursorComponent() {
    this.showCursor = true;
    this.timer =
        new Timer(
            500,
            e -> {
              showCursor = !showCursor;
              repaint();
            });
    this.timer.start();
  }

  @Override
  protected void paintComponent(Graphics g) {
    super.paintComponent(g);

    if (showCursor) {
      g.setFont(new Font("Monospaced", Font.PLAIN, 12));
      g.drawString("█", 10, 20);
    }
  }

  @Override
  public Dimension getPreferredSize() {
    return new Dimension(30, 30);
  }
}

File: src/main/kotlin/com/sourcegraph/cody/chat/ui/BlinkingCursorComponent.kt
package com.sourcegraph.cody.chat.ui

import com.intellij.openapi.Disposable
import com.intellij.util.ui.UIUtil
import java.awt.Dimension
import java.awt.Font
import java.awt.Graphics
import javax.swing.JPanel
import javax.swing.Timer

class BlinkingCursorComponent : JPanel(), Disposable {
  private var showCursor = true

  private val timer: Timer =
      Timer(500) {
        showCursor = !showCursor
        repaint()
      }

  init {
    timer.start()
  }

  override fun paintComponent(g: Graphics) {
    super.paintComponent(g)
    if (showCursor) {
      g.font = Font("Monospaced", Font.PLAIN, 12)
      g.drawString("█", 10, 20)
      g.color = UIUtil.getActiveTextColor()
      background = UIUtil.getPanelBackground()
    }
  }

  override fun getPreferredSize(): Dimension {
    return Dimension(30, 30)
  }

  override fun dispose() {
    timer.stop()
  }
}
